### PR TITLE
Prow presubmit job to build  mlkube.io container.

### PR DESCRIPTION
**Pull request by jlewi**
* This job is supposed to build a Docker container with the CRD. Once
  this works we will add another job to actually run the tests.
* We create a new image to be used by this job.
  * The image needs docker so that we can do a Docker build.

### CODE REVIEW

**Comment by jlewi** [comment]
@krzyzacy @foxish PTAL.

File: prow/config.yaml [hunk 1]
```diff
@@ -1821,6 +1821,41 @@ presubmits:
...
+          volumeMounts:
+          - name: service
+            mountPath: /etc/service-account
+            readOnly: true
+          - name: docker
+            mountPath: /var/run/docker.sock
+          ports:
+          # TODO(jlewi): Do we need this port?
+          - containerPort: 9999
+            hostPort: 9999
```
Comment (krzyzacy) [inline prow/config.yaml]: the port is for node affinity since they shares a git cache, you probably don't need it here.

File: prow/config.yaml [hunk 2]
```diff
@@ -1821,6 +1821,41 @@ presubmits:
...
+      trigger: "(?m)^/test( all| mlkube-build),?(\\s+|$)"
+      spec:
+        containers:
+        - image: gcr.io/mlkube-testing/builder:v20171010-352a232c-dirty
+          # TODO(jlewi): Add a service account to push to GCR?
+          # args:
+          env:
+            - name: GOOGLE_APPLICATION_CREDENTIALS
+              value: /etc/service-account/service-account.json
+          # TODO(jlewi): Do we need privileged mode in order to run Docker?
```
Comment (krzyzacy) [inline prow/config.yaml]: you probably need to figure out https://github.com/kubernetes/kubernetes/issues/1806, have you tried locally?

File: prow/config.yaml [hunk 3]
```diff
@@ -1821,6 +1821,41 @@ presubmits:
...
+          # TODO(jlewi): Do we need this port?
+          - containerPort: 9999
+            hostPort: 9999
+        volumes:
+        - name: service
+          secret:
+            secretName: service-account
+        - name: docker
+          hostPath:
+            path: /var/run/docker.sock
```
Comment (BenTheElder) [inline prow/config.yaml]: /cc @ixdy, I think we don't intend to allow jobs access to the host docker socket.

File: prow/config.yaml (hunk 2 above)
Comment (jlewi) [reply prow/config.yaml]: I've tried running my container locally using docker run but not running on K8s. Going to try it on my test prow cluster soon.

File: prow/config.yaml (hunk 3 above)
Comment (jlewi) [reply prow/config.yaml]: @BenTheElder Can you suggest an alternative approach?

Comment (krzyzacy) [reply prow/config.yaml]: @jlewi what I've been doing is using bazel to build/push the image instead, not sure if that's feasible to you

**Comment by jlewi** [comment]
@krzyzacy This is ready for another look but there are are some test failures I could use help with.

verify_test  is complaining that there is no test-grid associated with my presubmit job. 

//jobs:config_test is failing complaining about duplicate projects. I added an entry here because otherwise the //prow/config:go_default_test  failed complaining

```
jobs_test.go:152: Cannot find test-infra/jobs/mlkube-build-presubmit.sh for mlkube-build-presubmit
```
but it looks like my fix might have caused other problems.

File: jobs/config.json [hunk 4]
```diff
@@ -9907,6 +9907,27 @@
       "UNKNOWN"
     ]
   },
+  "mlkube-build-presubmit": {
```
Comment (krzyzacy) [inline jobs/config.json]: you'll need an equivalent one for `mlkube-build-postsubmit`

File: jobs/config.json [hunk 5]
```diff
@@ -9907,6 +9907,27 @@
...
+    "args": [
+      "--charts",
+      "--check-leaked-resources",
+      "--deployment=gke",
+      "--extract=gke",
+      "--gcp-cloud-sdk=gs://cloud-sdk-testing/ci/staging",
+      "--gcp-node-image=gci",
+      "--gcp-project=mlkube-testing",
```
Comment (krzyzacy) [inline jobs/config.json]: for the presubmit error you need to add something to https://github.com/kubernetes/test-infra/blob/master/jobs/config_test.py#L681-L791

**Comment by krzyzacy** [comment]
I think we don't want to allow sock mapping on prow pods. We should either use bazel to build docker images, or we need to implement some docker-in-docker service in prow.

**Comment by jlewi** [comment]
Thanks. I think I'll switch to using GCB to build the Docker images as a way around using bazel and mapping in the socket.

**Comment by krzyzacy** [comment]
@jlewi yeah, I think any daemonless build will work. We'll need to design a docker workflow solution in prow.

File: testgrid/config/config.yaml [hunk 6]
```diff
@@ -1537,6 +1537,10 @@ test_groups:
   gcs_prefix: kubernetes-jenkins/logs/ci-kubernetes-e2e-gke-gci-new-gci-master-upgrade-cluster-new
 - name: ci-kubernetes-e2e-mlkube-gke
   gcs_prefix: kubernetes-jenkins/logs/ci-kubernetes-e2e-mlkube-gke
+- name: mlkube-build-presubmit
+  gcs_prefix: kubernetes-jenkins/logs/mlkube-build-presubmit
+- name: mlkube-build-postsubmit
+  gcs_prefix: kubernetes-jenkins/logs/mlkube-build-postsubmit
```
Comment (krzyzacy) [inline testgrid/config/config.yaml]: also add to https://github.com/jlewi/test-infra/blob/8563b0c1fa76dfc890d05f38e6b9bb2faa5f76e6/testgrid/config/config.yaml#L3087-L3092

**Comment by krzyzacy** [comment]
hit by https://github.com/kubernetes/test-infra/issues/4989

**Comment by jlewi** [comment]
@krzyzacy  This is ready for another review.

- Only test failure appears to be due to #4989 
- I made the existing periodic test that @foxish had added consistent with the new pre/post submit tests
  - I created dashboards for each test on the bigdata dashboard group
  - Before it was using both big-data and apps
- I removed the use of docker build and used GCR instead to build the image
  - As discussed offline building the images with bazel might be better but I didn't want to introduce
    bazel in the mlkube toolchain just for this.

File: jobs/config_test.py [hunk 7]
```diff
@@ -789,6 +789,13 @@ def test_all_project_are_unique(self):
             'ci-kubernetes-node-kubelet-conformance': 'ci-kubernetes-node-kubelet-*',
             'ci-kubernetes-node-kubelet-benchmark': 'ci-kubernetes-node-kubelet-*',
             'ci-kubernetes-node-kubelet': 'ci-kubernetes-node-kubelet-*',
+
+            # The mlkube projects intentionally share projects,
+            # We map each test name to the same value so that below when we check for number of tests
+            # using this project there will just be one entry.
+            'mlkube-build-presubmit': 'mlkube-*',
+            'mlkube-build-postsubmit': 'mlkube-*',
+            'ci-kubernetes-e2e-mlkube-gke': 'mlkube-*',
```
Comment (krzyzacy) [inline jobs/config_test.py]: now you probably don't need this block. The test basically scans duplicated projects from config.json.

File: testgrid/config/config.yaml [hunk 8]
```diff
@@ -3540,9 +3542,15 @@ dashboards:
...
+    test_group_name: mlkube-build-periodic
+  - name: mlkube-build-postsubmit
+    description: Postsubmit tests of TfJob CRD running on stable GKE.
+    test_group_name: mlkube-build-postsubmit
+  - name: mlkube-build-presubmit
+    description: Presubmit tests of TfJob CRD running on stable GKE.
+    test_group_name: mlkube-build-presubmit
   notifications:
   - summary: Please configure this skeleton dashboard and remove this notification
     context_link: https://github.com/kubernetes/test-infra/tree/master/testgrid/config
```
Comment (krzyzacy) [inline testgrid/config/config.yaml]: (probably can delete this notification now)

Comment (jlewi) [reply testgrid/config/config.yaml]: Done

File: jobs/config_test.py (hunk 7 above)
Comment (jlewi) [reply jobs/config_test.py]: Done.

**Comment by jlewi** [comment]
Thanks.

File: prow/config.yaml [hunk 9]
```diff
@@ -3263,6 +3288,29 @@ postsubmits:
...
+        - name: service
+          mountPath: /etc/service-account
+          readOnly: true
+      volumes:
+      - name: service
+        secret:
+          secretName: service-account
+      - name: docker
+        hostPath:
+          path: /var/run/docker.sock
```
Comment (krzyzacy) [inline prow/config.yaml]: remove the sock map here as well?

File: prow/config.yaml [hunk 10]
```diff
@@ -2489,6 +2489,31 @@ presubmits:
         hostPath:
           path: /mnt/disks/ssd0
 
+  jlewi/mlkube.io:
```
Comment (krzyzacy) [inline prow/config.yaml]: hummm, I feel like our bot might not have access to your repo, you might want to configure somewhere around https://github.com/kubernetes/test-infra/blob/master/prow/plugins.yaml#L5-L11

Comment (krzyzacy) [reply prow/config.yaml]: also https://github.com/kubernetes/test-infra/blob/master/prow/plugins.yaml#L37-L38

Comment (jlewi) [reply prow/config.yaml]: Thanks. I also think I need to configure webhooks on my repo but I'll do that after this PR is done.

File: prow/config.yaml (hunk 9 above)
Comment (jlewi) [reply prow/config.yaml]: Done

**Comment by krzyzacy** [comment]
it's a good start and probably we have more stuff need to configure. we'll see how this goes :-)

/lgtm

**Comment by jlewi** [comment]
Thanks

**Comment by krzyzacy** [comment]
(whoops, forgot to ask you to squash :joy: )

**Pull request merged**